src/main/java/io/cryostat/rules/RuleMatcher.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.rules;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.script.Bindings;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+
+import io.cryostat.platform.ServiceRef;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import jdk.jfr.Category;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+
+class RuleMatcher {
+
+    private final ScriptEngine scriptEngine = new ScriptEngineManager().getEngineByName("nashorn");
+
+    public boolean applies(Rule rule, ServiceRef serviceRef) throws ScriptException {
+        RuleAppliesEvent evt = new RuleAppliesEvent(rule.getName());
+        try {
+            evt.begin();
+            Object result =
+                    this.scriptEngine.eval(rule.getMatchExpression(), createBindings(serviceRef));
```

Comment:
Why does `createBindings()` need to be called each time the script engine wants to compare a `matchExpression` to the `serviceRef`'s annotations?

Reply:
That's a good question/observation. The bindings need to be created to take the full `ServiceRef` structure and slightly simplify it before passing it into the script engine (ie turn `AnnotationKeys` just into Strings), and both the `matchExpression` and the `serviceRef` have constant state that doesn't change over time, so the `Bindings` object would be a good candidate for caching/memoization. I'm not sure how expensive it actually is to create the bindings however, so it may not be worth the overhead of adding a cache and dealing with cache invalidation when `ServiceRefs` disappear. The `BindingsCreationEvent` could help to profile this and figure out if it's worthwhile.

Reply:
We could also simply cache the result of whether a given rule's `matchExpression` applies to a given `ServiceRef`, instead of caching the `Binding`, and make it even faster by bypassing the script engine entirely, actually. If it is worth doing any caching then this would be the smarter way to go about it. This still incurs extra overhead in needing to do cache invalidation when `Rule` definitions are deleted or `ServiceRef`s disappear, though.

Reply:
... but then again, how often do we actually re-check a Rule against a ServiceRef? When a new Rule is added that rule alone is checked against all known ServiceRefs at the time, and vice-versa when a new ServiceRef appears it is checked against all of the Rules. So in either scenario, these are all new/unique combinations of Rule/ServiceRef, so the cache would be empty for all of these. Would we end up needing to re-evaluate it?

So maybe caching the ServiceRef -> Binding does make the most sense, since at least that would help in the case of a new Rule appearing. After the first one, we already have the Binding for each ServiceRef (assuming the set of ServiceRefs hasn't changed), so we can pass the same binding to the script engine to evaluate against the new rule matchExpression.

Reply:
Anyway, tl;dr, my hunch is that the majority of the time is spent actually setting up the script engine, passing data to it, executing it, and marshalling the results back into Java, and comparatively little time is spent creating the Binding object. If that's true then it probably isn't worth caching the Bindings.